Add/playlist block 805

antpb opened the pull request:
## Description
The playlist block matches the current functionality of Core's playlist shortcode. We are using the `ServerSideRender` component to display in the editor and are exposing settings to the editor for the different shortcode options. On each change, the block is updated.


<img width="287" alt="40888512-ded4d4c2-671d-11e8-9b36-94dfa1601584" src="https://user-images.githubusercontent.com/8985705/40933592-8f0ccbae-67f7-11e8-87a0-4397889532ec.png">

antpb committed:
initial commit

initial commit

antpb committed:
playlist init

starting playlist over

antpb committed:
storing ids in comment as ids attribute. moving to render callback....may need core mod >_<

storing ids in comment as ids attribute. moving to render callback....may need core mod >_<

antpb committed:
storing ids in comment as ids attribute. moving to render callback....may need core mod >_<

storing ids in comment as ids attribute. moving to render callback....may need core mod >_<

antpb committed:
playlist_block_callback_dependencies function. adds index.php

Commit:
cleaning up attributes and logging code. removes mediaelement component

danielbachhuber commented:
> The edit state currently does not render. I am working out that bit.

@antpb Have you looked at the `ServerSideRender` component? https://github.com/WordPress/gutenberg/blob/master/docs/blocks/creating-dynamic-blocks.md#live-rendering-in-gutenberg-editor

Commit:
serversiderender component being used now in edit...workign through attributes error

antpb commented:
thanks @danielbachhuber ! I see now how this needs to work. Working through an attribute error now. `Error loading block: Invalid parameter(s): attributes`

console is alerting: 
`GET http://local.wordpress.test/index.php?rest_route=/gutenberg/v1/block-renderer/core/playlist&context=edit&attributes%5Bids%5D=55%2C54 400 (Bad Request)` 

Currently looking into why that is. I've pushed what I have so far to this PR.

Commit:
removes client-asset changes meant for mediaelement component

Commit:
adds media-upload changes to support playlist modal

Commit:
adds type attribute for video/audio conditional for the render callback

joemcgill commented on core-blocks/playlist/index.js:
```diff
@@ -0,0 +1,167 @@
+		}
+
+		render() {
+			// const { registerBlockType } = wp.blocks;
```
Is this leftover from testing?

antpb replied:
I cleaned this up a bunch. Thanks! My PR is updated. Continuing to plug away at the edit state issue with the ServerSideRender component

antpb committed:
fixed edit state and now when you edit post, it acts as expected - commits from airport terminals ✈

antpb commented:
This may be the bug in my `ServerSideRender` problem: https://github.com/WordPress/gutenberg/pull/7232

investigating...

antpb committed:
Adjusted some of the wording to include video language...not just audio

antpb committed:
Group coding by antpb - noisysocks - soean

antpb committed:
Fixing noisysocks suggested changes

joemcgill committed:
Load playlist block alphabetically

This simply moves the playlist block in the list of blocks imported and registered in core-blocks/index.js.

joemcgill committed:
Clean up lint errors

This commit generally accomplishes a few thing:
- Removes any unused variables and functions
- Ensuring all used functions and variables are declared
- Simplifies some variable declarations
- Fixing spacing to conform to WP style

joemcgill committed:
Clean up dynamic rendering PHP code

This fixes up code style issues, adds correct docblocks, and removes unneccessary script rendering which isn't necessary during server rendering.

joemcgill committed:
Wrap server rendered playlists in figure element

This wraps the shortcode output in a `figure` element with a class of `wp-block-playlist` applied so we can style the block accordingly.

joemcgill committed:
Remove non-playlist styles from the playlist stylesheet.

joemcgill committed:
Fix playlist margins in the editor

This adds an editor.scss file to the playlist block that removes margins from `.wp-playlist-block` elements, to avoid margins being added as a default style by browsers.

joemcgill committed:
Fix alignment support for the playlist block

This adds an align attribute to the playlist block and adds the proper alignment class to the markup on the front end to support alignments. This also deletes the style.scss file, which now seems unnecessary.

antpb committed:
Merge pull request #1 from joemcgill/add/playlist-block

Playlist block updates @joemcgill  fixed a ton of formatting and bugs. :D

antpb committed:
Fixed edit button to auto populate media modal

antpb commented:
Thank you so much @joemcgill for the edits to this PR. Lots of lessons learned :D 

I've fixed the issue you mentioned about auto populating the media modal when clicking the edit button. We should be good to go now! @noisysocks is next step to give it a once over before merge? :)

antpb committed:
cleaned up spacing issues

Soean commented on core-blocks/playlist/edit.js:
```diff
@@ -0,0 +1,131 @@
+			<Fragment>
+				<BlockControls>
+					<Toolbar>
+					<MediaUpload
```
Indentation should be 6 tabs.

Soean commented on core-blocks/playlist/edit.js:
```diff
@@ -0,0 +1,131 @@
+		const { attributes, setAttributes, className } = this.props;
+		const { editing } = this.state;
+
+		const switchToEditing = () => {
```
`switchToEditing` is never used

Soean commented on core-blocks/playlist/index.php:
```diff
@@ -0,0 +1,46 @@
+	$attributes['ids'] = json_decode( $attributes['ids'] );
+
+	$classes = empty( $attributes['align'] ) ? 'wp-block-playlist' : 'wp-block-playlist align' . $attributes['align'];
+	$html = sprintf( '<figure class="%s">%s</figure>',
```
WPCS: Expect 4 spaces after `$html`.

Soean commented on core-blocks/playlist/index.php:
```diff
@@ -0,0 +1,46 @@
+	register_block_type( 'core/playlist', array(
+		'render_callback' => 'render_block_core_playlist',
+		'attributes'      => array(
+			'ids'  => array(
```
WPCS: 3 Spaces between `ids` and double arrow

Soean commented on core-blocks/playlist/index.php:
```diff
@@ -0,0 +1,46 @@
+			'ids'  => array(
+				'type' => 'string',
+			),
+			'type' => array(
```
WPCS: 2 Spaces between `type` and double arrow

Soean commented on core-blocks/playlist/index.js:
```diff
@@ -0,0 +1,43 @@
+		},
+	},
+
+	supports: {
```
Add `html: false` because there is no HTML view.

Soean reviewed (commented):
Great work! I added some comments to the code.

Soean commented:
- The Upload Button isn't working.
- The block is ignoring the playlist settings
<img width="227" alt="bildschirmfoto 2018-06-17 um 14 51 20" src="https://user-images.githubusercontent.com/695201/41508015-fb8bd2e8-723d-11e8-92d9-9319e92fd65e.png">

antpb committed:
fixes Soean's suggestions and adding attributes to fix settings issue

antpb commented:
thanks @Soean! I fixed most of those issues and pushed the changes out for more testing. Still pending is the playlist settings and upload button fixes. Having some issues getting the media modal to send back the settings options. I'm looking into playlist shortcode to understand how the attribute is being passed back. I've added the options to the render callback though so once I have a way to pass them from the modal to set attributes we'll be golden. If anyone has any idea how to do that, or a direction I should be looking please feel free to guide me :) will report back soon.

antpb committed:
removes some silly test code

Soean committed:
Add controls and fix lint errors

noisysocks commented:
@antpb: You can get the setting in `onUpdate` using e.g. `selectedImages.playlist.get( 'tracklist' )`. `'tracklist'` corresponds to the `data-setting` attribute on the setting's checkbox. Note that an `undefined` value means that the setting is turned on. 

Having said that, I think for Gutenberg, we should just not show those settings in the media picker and instead allow them to be configured in the block inspector.

noisysocks commented:
Thanks so much your hard work on this! You took on a challenging block for your first contribution 😍 

In the medium to long term, I'd like to see us move away from using `ServerSideRender` and building out an `edit` UI that is more interactive, e.g. lets users re-arrange and remove tracks within the editor. This would align better with Gutenberg's goal of allowing direct manipulation. 

But this is a great start, and I'm stoked to get this in as an MVP that we can iterate on.

From my testing, it's looking really close. Just a few things:

1. It looks like this breaks the Gallery block. I'm no longer able to select multiple images when creating a gallery in Gutenberg.
2. As above, I think let's just remove the 'Playlist Settings' sidebar from the media picker. This should be possible by removing `SettingsView: wp.media.view.Settings.Playlist` from `media-upload/index.js`.
3. Can we reduce the block padding so that, when the block is selected in the editor, the playlist appears flush against the block's outline?

Pinging @karmatosed for any design feedback she might have. I will properly review the code once we're in a good place functionally.

antpb committed:
Merge branch 'master' into add/playlist-block-805

karmatosed commented:
I wanted to give this a test but don't seem to get it able to run without errors. Sorry about that could you try and check if it's something my end of with the PR @antpb?

antpb commented:
@karmatosed happily! I have time slated for later today to get the final bugs here solved. I'll ping you when that is done. Maybe a lofty goal but I am hoping to get this ready to go before our next release in a few days. I don't foresee a ton of struggle getting the fixes @noisysocks mentioned above resolved. :)

antpb committed:
removes `SettingsView` from Playlist Media view

antpb committed:
fixes gallery mutiple bug - Nice find @noisysocks !

antpb committed:
sets defaults to on for tracklist, images, artist as per Core defaults

antpb committed:
sets defaults for front end render as well as fixes lint errors (someone finally installed a linter!)

antpb commented:
@noisysocks thanks for the feedback. It was so nice to have a concise list of things to do. :) I removed the playlist `SettingsView`, but backbone reaaaally didn't like it. It technically works but I don't like the error. Is there a way to set `SettingsView` to false instead of omitting it that you know of?

Error: Uncaught TypeError: g is not a constructor

```        renderSettings: function(a) {
            var b = this.get("library")
              , c = this.get("collectionType")
              , d = this.get("dragInfoText")
              , g = this.get("SettingsView")
              , h = {};
            b && a && (b[c] = b[c] || new Backbone.Model,
            h[c] = new g({
                controller: this,
                model: b[c],
                priority: 40
            }),
```

I think this may require some kind of condition added in core to allow `SettingsView` to be omitted. Looking into that and the upload button still, but a majority of the bugs you've identified are now fixed. I've gone ahead and also set the defaults to true on our playlist options as this is the expected behavior in core.

I found another bug in the settings — on initial creation of the block, the attributes are not saved for `tracklist, images, artists` but once you edit the post and check the options, it saves without fail. Weird one, but probably something silly. Digging continues!

Soean commented on core-blocks/playlist/edit.js:
```diff
@@ -0,0 +1,162 @@
+
+		const setAudio = ( [ audio ] ) => onSelectMedia( audio );
+
+		const uploadFromFiles = ( event ) => editorMediaUpload( event.target.files, setAudio, 'audio' );
```
The only parameter of `editorMediaUpload` should be an object, I think this can solve the upload problem.

antpb replied:
maybe a simple fix....considering the existing behavior of the editor is to `add media` and then click playlist>upload, do we really need an upload button? On initial click of the media button, you have the option to upload in the first tab @karmatosed would you agree? The upload button was copied from a early iteration of audio

antpb replied:
btw, still working to get this fixed so we can have the option. Just want to verify that it is actually a wanted feature of the block

Soean commented on core-blocks/playlist/edit.js:
```diff
@@ -0,0 +1,162 @@
+		const onSelectMedia = ( media ) => {
+			//check if there are returned media items and set attributes when there are
+			if ( media && media[ 0 ].url ) {
+				media = ( 1 < media.length ) ? media : [ media ];
```
What does this line? If you select just one file, you create an array of an array?

antpb replied:
haha yeah I guess it does. This serves no function. I've removed it and validated that selecting one item works as we would expect. Thanks!

Soean commented on core-blocks/playlist/edit.js:
```diff
@@ -0,0 +1,162 @@
+
+		if ( editing ) {
+			return (
+				<Placeholder
```
Maybe we should use the `MediaPlaceholder` here?

antpb committed:
removes silly array of arrays and adds back in SettingsView to make backbone happy

antpb committed:
adds style control for playlist: dark, light

antpb committed:
should fix merge conflict

antpb committed:
changes edit-post/hooks/blocks to components for merge conflict

antpb committed:
fixes more merge conflicts

antpb committed:
Merge branch 'master' into add/playlist-block-805

antpb commented:
Just chiming in here with an update. I've started work on fixing the upload button and the following commit outlines my attempts. [c21a650 ](https://github.com/WordPress/gutenberg/pull/7126/commits/c21a65000ab5a3d39ca4a6803c5161746541d282) I'm still having issues with this one but am continuing to dig. 

To take my mind off the issues I've been having with Backbone/`SettingsView` and Upload Button, I threw in an extra feature for this block. I added style controls in the settings so users can change to the inverted color scheme. Options are 'light' and 'dark' with default set to light to match Core. This is a shortcode option previously unexposed to users in the editor, so that's kinda cool. 

Moving my focus back on the evil bugs. :)

antpb committed:
uses new IconButton component to match gallery

antpb committed:
Merge branch 'add/playlist-block-805' of https://github.com/antpb/gutenberg into add/playlist-block-805

antpb committed:
fixes `ServerSideRender` update for settings

antpb committed:
@soean provided me the golden snippet of code to fix uploads

antpb committed:
removes `SettingsView`

antpb committed:
linting fix from spaces to tabs

Soean committed:
Fixes WPCS warnings

Soean committed:
Remove isPlainObject and map from SSR

antpb committed:
fixes `Media Library` button

antpb commented:
With @Soean's massive contributions here I think this is ready for design review @karmatosed :) 

FYI Backbone didnt like us removing `SettingsView` from the playlist views so there currently is a console error. If you are having issues getting through the flow of selecting tracks, let me know and I'll revert the change. (or maybe later I'll find a way to fix)

antpb committed:
auto sets type based on first item in media array passed back from uploads and  media selection

antpb committed:
adds type attribute to edit button to pass back to the media views...need to add condition in media view for type

antpb committed:
adds edit condition for videos so media view focuses on video's and not a default of audio

antpb committed:
changes frame to select since that is what a user would probably want to do

karmatosed commented:
This is great work! Thanks for working through this @antpb. I am going to give a few bits of design feedback so hopefully we can wrap up things for you.

Can you remove the border so it just sits without one? Also balance up the padding on the top and bottom.

![2018-06-22 at 21 55](https://user-images.githubusercontent.com/253067/41798851-1f1bafc6-7667-11e8-8b97-255b82345af6.png)

You can see it better here:

![2018-06-22 at 21 57](https://user-images.githubusercontent.com/253067/41798877-46a4dd6a-7667-11e8-8755-85603594db20.png)

I am not sure what the setting 'Show Artist Name in Tracklist' is doing as doesn't seem to work for me.

karmatosed reviewed (changes requested):
Just a few little changes as mentioned above.

antpb commented:
no problem @karmatosed ! Thanks so much for the speedy design review! I'll have this patched up within the next day or two. :)

antpb committed:
adds @karmatosed design recommendations and fixes an upload bug introduced

when uploading, we are assuming that the type is audio by default because for whatever reason the upload form is not passing back the mime type

antpb commented:
hi @karmatosed ! I've added the design feedback changes to this as well as fixed an issue with the upload button. 

Two minor bugs pending. Not sure if this is a blocker to merge and iterate but:

In the `addFiles` function we automatically assume `audio` to be the type set in attributes because for whatever reason the upload form is not sending back the type. I was attempting to use the first item in the `media` array to pull type but that is not returning anything. This approach worked fine with the `onSelectMedia` function. I'm thinking the solution here would be to take the first object in `ids` and use that to query the type in the `initializePlaylist` function. I don't feel like that is the best approach but I think that would be the right time to fire `setAttribute` for type.

Also, I've added back in the `SettingView` for playlists so we make Backbone happy for the time being. We may need to take @noisysocks recommendation on getting back the settings from the media modal. A bit of thinking needed on this one..

antpb commented:
Oh and the `Show Artist Name in Tracklist` bug mentioned  I think the reason you were not seeing the artist name do anything is because the track name of the one that has an artist name in your example is too long to show the artist name appearing/disappearing. The track after that one doesn't seem to have an artist name set in the meta.

Here are some before/after screenshots to illustrate the option. :)

![screen shot 2018-06-25 at 11 14 51 am](https://user-images.githubusercontent.com/8985705/41862264-09a101c4-7869-11e8-81e2-5614284a5f2e.png)
![screen shot 2018-06-25 at 11 14 43 am](https://user-images.githubusercontent.com/8985705/41862265-09b2cd5a-7869-11e8-819b-f6918e5d7495.png)

Soean committed:
Improve toggle functions

antpb committed:
hides settings in media modal on gutenberg editor pages. ¯\_(ツ)_/¯

Soean committed:
Fix toggle style

antpb committed:
fixing merge conflict

antpb committed:
fixing merge issue of including package-lock

antpb committed:
updating package-lock to match master because my local environment was silly

antpb committed:
Merge branch 'master' into add/playlist-block-805

antpb committed:
final changes to remove package-lock.json from this PR

antpb committed:
removing extra =

Soean committed:
Alow upload video and audio files

Soean commented:
In my last commit I added `mime_type` to the return object of uploaded files. 
Now it's also allowed to define more types in allowedType for uploads. Maybe we change it to `allowedTypes` and only accept arrays?

I think we fixed all known bugs, now we can optimize the code.

Soean commented on core-blocks/playlist/index.js:
```diff
@@ -0,0 +1,60 @@
+
+	description: __( 'The Playlist block allows you to embed playlist files and play them back using a Core playlist.' ),
+
+	icon: 'format-audio',
```
Should we change the icon to `dashicons-playlist-video`?
https://developer.wordpress.org/resource/dashicons/#playlist-video

antpb replied:
I love the idea of that. I struggled to decided an icon and the only reason I went with the audio is because technically both video and audio can produce sound. haha

`dashicons-playlist-video` makes so much more sense. Great suggestion!

Soean committed:
Refactor Code

Soean committed:
Add tests

antpb committed:
changes playlist icons to better indicate both video AND audio.

antpb committed:
Revert "changes playlist icons to better indicate both video AND audio."

This reverts commit 4495b53d1375a4be9df3dffb0339526f0a4e58b3.

karmatosed reviewed (approved):
I think the design works so let's get this in as approved design wise. Assuming there is a final code review to happen let's get that happen or lets merge. Thanks for all the amazing hard work on this.

noisysocks reviewed (changes requested):
This is looking really close! Keep up the great work—I reckon we can get this merged this week 🙂

antpb commented on core-blocks/playlist/edit.js:
```diff
@@ -0,0 +1,188 @@
+			filesList: files,
+			onFileChange: ( media ) => {
+				if ( media && media[ 0 ].mime_type ) {
+					const type = media[ 0 ].mime_type.split( '/' )[ 0 ];
```
The flow of the media library modal will not allow more than one type considering you have to select audio playlist or video playlist. The problem would occur when someone uploads video and audio together using the upload button. We could have two upload buttons that have type restrictions. Thinking `Upload Video` / `Upload Audio` Here's a screenshot of what I'm thinking. @karmatosed what do you think? From a design perspective does this feel weird? I feel like it's odd. 
![screen shot 2018-07-03 at 11 29 37 am](https://user-images.githubusercontent.com/8985705/42232596-86af039a-7eb4-11e8-90ab-d456bfe5263a.png)

noisysocks replied:
Agree that that looks a bit odd.

Just spitballing: Could we introduce an error notice, e.g. if you upload audio and video files together there is a message that appears to the affect of _Error! Please select only video files or only audio files._ We have `withNotices` which is a convenient way of implementing this sort of UI.

antpb replied:
I'm trying to think of the best way to validate the types. My first pass at this is to use lodash `uniq` but what I'm struggling with is finding a length for that. My thoughts are that we map through `media` in `onFileChange` and make an array of the unique `mimeType`. If unique types is greater than 1, we throw notice. Is this thinking right @noisysocks ?

noisysocks replied:
Yep, that approach would work.

```js
const areTypesConsistent = uniqBy( media, media => media.mimeType ).length === 1;
if ( ! areTypesConsistent ) {
	// show an error
}
```

You could also use something like [`_.every`](https://lodash.com/docs/#every) to check that all types match the first type.

```js
const firstType = get( media, [ 0, 'mimeType' ] );
const areTypesConsistent = !! firstType && every( media, ( media ) => media.mimeType === firstType );
if ( ! areTypesConsistent ) {
	// show an error
}
```

antpb commented on core-blocks/playlist/edit.js:
```diff
@@ -0,0 +1,188 @@
+			//check if there are returned media items and set attributes when there are
+			if ( media && media[ 0 ].url ) {
+				const ids = JSON.stringify( media.map( ( item ) => item.id ) );
+				setAttributes( { tracklist, artists, images, ids, type: media[ 0 ].type } );
```
I think `onSelectMedia` is a pretty clear name of the function. We could just remove the current one and rename `addFiles` to `onSelectMedia`

noisysocks replied:
We're admittedly very inconsistent with this, but I believe the current consensus within Gutenberg is to name functions (including `onEvent` callbacks) so that they

a) Begin with a verb
b) Describe _what_ the callback does, rather than what event it handles

antpb replied:
I went ahead and renamed `addFiles` to `onUploadFiles` to fit the naming convention. After comparing to `onSelectMedia` I remembered why they were separate. One of these handles the objects returned from the browser upload and the other handles the objects from selecting items from the Media modal. There might be a better way to do this but for the time being I'll at least get this named proper. Feel free to change this if you'd like. :) 
Also removed `tracklist`, `artists`, and `images` from `setAttributes` Those were there in anticipation of figuring out how to pass the settings back from the media modal. Now we hide the settings, so no need. Fixed and moving to other changes.

antpb commented on core-blocks/playlist/index.php:
```diff
@@ -0,0 +1,62 @@
+ * @return string Playlist output. Empty string if the passed type is unsupported.
+ */
+function render_block_core_playlist( $attributes ) {
+	$attributes['ids'] = json_decode( $attributes['ids'] );
```
Yeah my memory is a little hazy on the reasons behind this but I remember it being due to an error thrown by the REST API.  I'll look into it and document if I can find why this was.

noisysocks replied:
From memory I think it's because Gutenberg isn't clever enough to parse/serialise the array as `ids[]=1&ids[]=2&ids[]=3` which is what PHP expects and generates.

antpb commented on core-blocks/playlist/test/index.js:
```diff
@@ -0,0 +1,13 @@
+	test( 'block edit matches snapshot', () => {
+		const wrapper = blockEditRender( name, settings );
+
+		expect( wrapper ).toMatchSnapshot();
```
woo! thanks @Soean for building this!!

antpb commented on lib/client-assets.php:
```diff
@@ -1233,6 +1233,8 @@ function gutenberg_editor_scripts_and_styles( $hook ) {
 	) );
 	wp_enqueue_editor();
+	wp_playlist_scripts( 'audio' );
+	wp_playlist_scripts( 'video' );
```
I remember it was something like 110kb - 120kb

antpb commented on core-blocks/playlist/edit.js:
```diff
@@ -0,0 +1,188 @@
+					</FormFileUpload>
+					<MediaUpload
+						onSelect={ onSelectMedia }
+						type="audio"
```
When you click the Media Library button we need a default type to set for when the modal is opened. I set this to `audio` just because it was the first option of the two. I have no data to support this, but my gut tells me playlists are used more often for Audio than Video so the default might be fair. Thoughts?

antpb replied:
yeah that makes much more sense. fixed

antpb commented on core-blocks/playlist/edit.js:
```diff
@@ -0,0 +1,188 @@
+			}
+		};
+
+		const mediaIds = this.props.attributes.ids && this.props.attributes.ids.replace( /^\[(.+)\]$/, '$1' ).split( ',' );
```
for whatever reason I was having issues using `JSON.parse` I'll try again and see what I can find.

antpb committed:
refactor naming of editing state and code tweak

antpb commented on core-blocks/playlist/edit.js:
```diff
@@ -0,0 +1,188 @@
+
+		// check for if ids is set to determine edit state
+		this.state = {
+			editing: ! this.props.attributes.ids,
```
agreed. fixed.

antpb committed:
fixes conditional in onFileChange

antpb commented on core-blocks/playlist/edit.js:
```diff
@@ -0,0 +1,188 @@
+			allowedType: [ 'audio', 'video' ],
+			filesList: files,
+			onFileChange: ( media ) => {
+				if ( media && media[ 0 ].mime_type ) {
```
yep yep! makes more sense to me. fixed

antpb committed:
refactors naming of addFiles to onUploadFiles and removes unnecessary attributes

antpb committed:
adds period to Placeholder instructions

antpb commented on core-blocks/playlist/edit.js:
```diff
@@ -0,0 +1,188 @@
+				<Placeholder
+					icon="media-audio"
+					label={ __( 'Audio/Video Playlist' ) }
+					instructions={ __( 'Select audio or video files from your library, or upload a new ones' ) }
```
nice catch! fixed :)

antpb committed:
fixes accepted types to comma-separated string

antpb commented on core-blocks/playlist/edit.js:
```diff
@@ -0,0 +1,188 @@
+						multiple
+						className="wp-block-playlist__upload-button"
+						onChange={ this.uploadFromFiles }
+						accept="audiovideo"
```
yep yep! Works nicely.

antpb commented on core-blocks/playlist/edit.js:
```diff
@@ -0,0 +1,188 @@
+							label={ __( 'Playlist Style' ) }
+							value={ style }
+							onChange={ this.toggleAttribute( 'style' ) }
+							options={ styleOptions }
```
agreed. Fixed.

antpb committed:
fixes style options to be inline. maybe more styles some day :D

antpb committed:
sets type default to audio to initialize media modal

antpb committed:
sets mediaUpload value correctly using `mediaIds`

antpb commented on core-blocks/playlist/edit.js:
```diff
@@ -0,0 +1,188 @@
+						type="audio"
+						multiple
+						playlist
+						value={ this.props.attributes.ids }
```
oh yeah! cool. Fixed.

antpb committed:
spaces child of component correctly

antpb commented on core-blocks/playlist/edit.js:
```diff
@@ -0,0 +1,188 @@
+								icon="edit"
+								label={ __( 'Media Library' ) }
+								onClick={ open }
+							> { __( 'Media Library' ) }
```
yep yep!

antpb commented on core-blocks/playlist/edit.js:
```diff
@@ -0,0 +1,188 @@
+			);
+		}
+
+		/* eslint-disable jsx-a11y/no-static-element-interactions, jsx-a11y/onclick-has-role, jsx-a11y/click-events-have-key-events */
```
I was matching the structure for the audio block from way back. We likely do not need these. I'm going to remove em.

antpb committed:
removes lint rules

antpb committed:
css spacing

antpb committed:
simplifies margin css

antpb committed:
renames artists attribute to better indicate action

antpb committed:
fixes typo

antpb committed:
fixes sneaky tab

antpb committed:
fixes other sneaky tab

antpb committed:
renames `playlistItemObject` to `slimPlaylistItemObject`

antpb committed:
fixes abbreviations

antpb commented on edit-post/hooks/components/media-upload/index.js:
```diff
@@ -140,6 +250,9 @@ class MediaUpload extends Component {
+			if ( playlist ) {
+				onSelect( selectedImages.models.map( ( model ) => playlistItemObject( model.toJSON() ) ) );
+			}
 			onSelect( selectedImages.models.map( ( model ) => this.processMediaCaption( slimImageObject( model.toJSON() ) ) ) );
```
yep that second one is the one we should go with. I had a lot of issues visualizing this in my head when initially writing this. Makes much more sense now.

antpb committed:
fixes strcuture of `multiple` and `playlist` checks in mediaupload

antpb committed:
fixes slimPlaylistItemObject

antpb commented on utils/mediaupload.js:
```diff
@@ -69,7 +69,10 @@ export function mediaUpload( {
-	const isAllowedType = ( fileType ) => startsWith( fileType, `${ allowedType }/` );
+	const isAllowedType = ( fileType ) => {
+		allowedType = [].concat( allowedType );
+		return allowedType.some( ( type ) => startsWith( fileType, `${ type }/` ) );
```
cool. I dont remember touching this for playlists but I went ahead and added `castArray` and refactored as suggested. We'll probably want to test on anything that depends on this or revert.  Feel free to make that call. :)

antpb committed:
adds castArray to utils/mediaupload

antpb commented on utils/mediaupload.js:
```diff
@@ -118,6 +121,7 @@ export function mediaUpload( {
 					caption: get( savedMedia, [ 'caption', 'raw' ], '' ),
 					id: savedMedia.id,
 					link: savedMedia.link,
+					mime_type: savedMedia.mime_type,
```
can certainly make the change, but I'm not sure how this will impact anything dependent. Does anything outside this consume `mime_type` as it is defined here?

antpb replied:
Oh, we can rename. this was added in for playlists. fixed.

antpb commented on utils/mediaupload.js:
```diff
@@ -43,13 +43,13 @@ export function getMimeTypesArray( wpMimeTypesObject ) {
+ * @param   {Function}     $0.onError           Function called when an error happens.
+ * @param   {Function}     $0.onFileChange      Function called each time a file or a temporary representation of the file is available.
  */
 export function mediaUpload( {
```
@Soean have any thoughts on this?

antpb commented on core-blocks/playlist/editor.scss:
```diff
@@ -0,0 +1,12 @@
+
+.gutenberg-editor-page .playlist-settings {
+	display: none;
+}
```
fixed.

antpb committed:
Merge branch 'master' into add/playlist-block-805

antpb committed:
fixes `showArtists` in index.php

antpb committed:
sets default for php render to default 'audio' fixes a bug introduced with defaults. :)

antpb committed:
adds space again to css file

antpb commented on core-blocks/playlist/index.js:
```diff
@@ -0,0 +1,60 @@
+	edit,
+
+	save() {
+		return null;
```
I don't think there is any support in feeds currently for playlists. I love the thinking here though. What would be the best course of action? I've been out of the loop on feed related stuff.

noisysocks replied:
Yeah, I just had a brief look and it doesn't seem that there's any page in WordPress that renders a playlist e.g. given its ID. 

Never mind 🙂

antpb committed:
fixes upload issue adding back in mime_type removed from merge conflict

antpb committed:
renames mime_type to mimeType

antpb committed:
removes test code for another issue...

antpb commented:
Alrighty! Out of time for the day, but I got 90% of the issues mentioned from @noisysocks' review addressed. I have one pending question about the notice logic for multi-type playlists: https://github.com/WordPress/gutenberg/pull/7126/files/ecda6d825bc45cab37c7bf4d2e73c9e6c78005f2#r200453130

Once that is decided/fixed, we will have just a few items pending that were mentioned in the review above. Some requests for review with folks more familiar with backbone and a decision to be made around testing.

karmatosed commented on core-blocks/playlist/index.js:
```diff
@@ -0,0 +1,60 @@
+export const settings = {
+	title: __( 'Playlist' ),
+
+	description: __( 'The Playlist block allows you to embed playlist files and play them back using a Core playlist.' ),
```
Yes let's not say Core.

antpb committed:
sets attribute 'artists' as 'showArtists' as the shortcodes expects. Also sets work for notice on mismatch type

antpb commented on core-blocks/playlist/index.js:
```diff
@@ -0,0 +1,60 @@
+		tracklist: {
+			type: 'boolean',
+			default: true,
+		},
```
I introduced a bug by renaming this. The shortcode expects an attribute `artists` so I've added a declaration of `artists` equal to `showArtists` in the render callback. I think that should fix it.

antpb committed:
adds showTrackNumbers attribute to expose more of the shortcode options

antpb committed:
uses this.props.attributes to fix edit state bug.

For some reason passing attributes to the constructor does not allow the edit state to be determined. reverting to this.props.attributes fixes the issue

antpb committed:
changes playlist block description to exclude `core` verbiage

aduth commented:
Should this be closed in favor of #9169 ?

antpb commented:
Closed. Work is being done in #9169
